### PR TITLE
feat: allow searching a datasource without a text query

### PR DESCRIPTION
**Pull request by fontanierh**
- make `query` and optional param when searching a data source in core
- uses the `core` SQL DB to retrieve documents ordered by timestamp desc (with the same filter options)
- then uses qdrant to fetch the points

### CODE REVIEW

File: core/src/stores/postgres.rs [hunk 1]
```diff
@@ -1330,6 +1332,98 @@ impl Store for PostgresStore {
...
+                + &(param_index + 1).to_string();
+            params.push(Box::new(limit as i64));
+            params.push(Box::new(offset as i64));
+        }
+        let params_ref: Vec<&(dyn ToSql + Sync)> = params.iter().map(|b| &**b).collect();
+
+        let query_future = c.query(&query, &params_ref);
+
+        // This line breaks
+        query_future.await?;
```
Comment (fontanierh) [inline core/src/stores/postgres.rs:1422]: I cannot await this because `params` is not `Send`

Comment (fontanierh) [reply core/src/stores/postgres.rs:1422]: @spolu

File: core/src/documents_search_filter.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
```
Comment (spolu) [inline core/src/documents_search_filter.rs:2]: Should be under `data_sources` ?

Comment (fontanierh) [reply core/src/documents_search_filter.rs]: Moved back 👍

File: core/src/data_sources/data_source.rs [hunk 3]
```diff
@@ -1266,6 +1166,130 @@ impl DataSource {
...
+        let (doc_ids, _) = store
+            .find_data_source_document_ids(
+                &self.project,
+                self.data_source_id(),
+                filter,
+                // with top_k documents, we should be guaranteed to have at
+                // least top_k chunks, if we make the assumption that each
+                // document has at least one chunk.
+                // TODO @fontanierh: confirm this is actually true
+                Some((top_k, 0)),
```
Comment (fontanierh) [inline core/src/data_sources/data_source.rs:1187]: @spolu  This is one assumption that may be untrue. If it is wrong, then in some unlucky cases we may end up with less chunks than top_k. 
But I don't really see how it's possible

File: core/src/data_sources/data_source.rs [hunk 4]
```diff
@@ -800,194 +802,81 @@ impl DataSource {
...
-                    }
-                    None => (),
-                };
-
-                Some(qdrant::Filter {
-                    must: must_filter,
-                    must_not: must_not_filter,
-                    ..Default::default()
-                })
+        let chunks = match query.is_none() {
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:808]: generally match query {
  Some(..) =>
  None => 
 }
 
 prevents you from unwraping in the is not none case

File: core/src/data_sources/data_source.rs [hunk 5]
```diff
@@ -800,194 +802,81 @@ impl DataSource {
...
+            true => {
+                let store = store.clone();
+                if !target_document_tokens.is_none() {
+                    Err(anyhow!(
+                        "target_document_tokens is only supported with query"
+                    ))?;
+                }
+                self.retrieve_chunks_without_query(store, &qdrant_client, top_k, &filter)
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:816]: Let's add a DSSTAT like log?

File: core/src/data_sources/data_source.rs (hunk 3 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:1187]: If "it works" even if we get some empty documents then it's fine :+1: can only be a comment to bring the reader attention to that.

**Review by spolu — CHANGES_REQUESTED** [review]
Minor changes, LGTM otherwise!

File: core/src/data_sources/data_source.rs (hunk 5 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs:816]: Would you add inside of the helper or around ?

Comment (spolu) [reply core/src/data_sources/data_source.rs:816]: outside since the start time is defined here?

File: core/src/data_sources/data_source.rs (hunk 4 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs]: Fixed

File: core/src/data_sources/data_source.rs (hunk 5 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs]: Done

File: core/src/data_sources/data_source.rs (hunk 3 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs:1193]: it does work 👍

File: core/src/data_sources/data_source.rs [hunk 6]
```diff
@@ -1266,6 +1172,122 @@ impl DataSource {
...
+                filter,
+                // with top_k documents, we should be guaranteed to have at
+                // least top_k chunks, if we make the assumption that each
+                // document has at least one chunk.
+                Some((top_k, 0)),
+            )
+            .await?;
+
+        let qdrant_batch_size: usize = 10; // number of `document_ids` to query at once in QDrant
+        let qdrant_page_size: u32 = 100; // number of points to fetch per page in QDrant
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:1197]: Let's use a power of 2 (128)

File: core/src/data_sources/data_source.rs [hunk 7]
```diff
@@ -1266,6 +1172,122 @@ impl DataSource {
...
+                self.data_source_id(),
+                filter,
+                // with top_k documents, we should be guaranteed to have at
+                // least top_k chunks, if we make the assumption that each
+                // document has at least one chunk.
+                Some((top_k, 0)),
+            )
+            .await?;
+
+        let qdrant_batch_size: usize = 10; // number of `document_ids` to query at once in QDrant
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:1196]: Let's use a power of 2 (8)

File: core/src/data_sources/data_source.rs [hunk 8]
```diff
@@ -1266,6 +1172,122 @@ impl DataSource {
...
+                ..Default::default()
+            };
+
+            let mut page_offset: Option<PointId> = None;
+            let mut batch_points: Vec<RetrievedPoint> = vec![];
+
+            // we must scroll through all result pages of a batch, because we want to
+            // to sort by reverse-chron timestamp and then by chunk_offset
+            // and QDrant doesn't support any kind of sorting
+            loop {
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:1234]: I get antsy with potential infinite looping, can we add a sentinel to defense ourselves against qdrant going loco?

Comment (spolu) [reply core/src/data_sources/data_source.rs:1234]: It's so ugly I agree, but infinite loops are worse?

**Review by spolu — DISMISSED** [review]
LGTM

File: core/src/data_sources/data_source.rs (hunk 8 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs:1234]: Yeah sounds good 👍

Comment (fontanierh) [reply core/src/data_sources/data_source.rs:1236]: as per IRL: went with 16k iterations max. If we do reach 16k iterations, then we'll return an error.

**Pull request merged**